daemon/migrator.c

```diff
@@ -0,0 +1,804 @@
+}
+
+/* ------------------------------------------------------------------------- *
+ * MIGRATOR_REMOVAL_QUEUE
+ * ------------------------------------------------------------------------- */
+
+static void
+migrator_removal_enqueue(migrator_t *self, const gchar *path)
+{
+    g_queue_push_tail(self->mgr_removal_queue, g_strdup(path));
+}
+
+static bool
+migrator_removal_queued(migrator_t *self)
+{
+    return self->mgr_removal_queue->length != 0;
+}
+
+static void
+migrator_removal_dequeue_all(migrator_t *self)
+{
+    for( GList *iter = self->mgr_removal_queue->head; iter; iter = iter->next ) {
+        gchar *path = iter->data;
+        iter->data = NULL;
+        remove_approval_path(path);
+        g_free(path);
+    }
+    g_queue_clear_full(self->mgr_removal_queue, NULL);
+}
+
+static void
+migrator_removal_clear_queue(migrator_t *self)
+{
+    g_queue_clear_full(self->mgr_removal_queue, g_free);
+}
+
+/* ------------------------------------------------------------------------- *
+ * MIGRATOR_PROCESS
+ * ------------------------------------------------------------------------- */
+
+static void
+migrator_process_file(migrator_t *self, const gchar *path)
+{
+    bool           migrated    = false;
+    gchar         *appname     = NULL;
+    appsettings_t *appsettings = NULL;
+    approval_t    *approval    = approval_create(self, path);
+    stringset_t   *granted     = stringset_create();
+
+    if( approval_valid(approval) ) {
+        const stringset_t *available = control_available_applications(migrator_control(self));
+
+        appname = path_to_desktop_name(approval_profile(approval));
+        if( stringset_has_item(available, appname) ) {
+            const appinfo_t *appinfo = migrator_appinfo(self, appname);
+
+            if( !g_strcmp0(approval_organization(approval), appinfo_get_organization_name(appinfo)) &&
+                !g_strcmp0(approval_application(approval), appinfo_get_application_name(appinfo)) ) {
+                /* TODO: The plugin stored also binary path in the
+                 *       approval files but currently that's just
+                 *       ignored here. However, we could check that we
+                 *       are approving the right binary if we can tell
+                 *       which path it's using.
+                 */
+
+                appsettings = settings_appsettings(migrator_settings(self), approval_uid(approval), appname);
+                stringset_extend(granted, appsettings_get_granted(appsettings));
+                stringset_extend(granted, approval_permissions(approval));
+                appsettings_set_granted(appsettings, granted);
+                appsettings_set_allowed(appsettings, APP_ALLOWED_ALWAYS);
+                migrated = true;
+                log_info("%s migrated", path);
+            }
+        }
+    }
+
+    if( !migrated )
+        log_warning("%s was not migrated", path);
+
+    migrator_removal_enqueue(self, path);
```

Comment:
This does ownership transfer of dynamically allocated string,  allocated in migrator_build_queue() and eventually freed in migrator_removal_dequeue_all() / migrator_removal_clear_queue(), right?

Which causes mentail disconnect/confusion in my head to have functions in between those two ends (migrator_process_file and migrator_removal_enqueue) to claim the data to be of 'const char *' type.

Reply:
There is `g_strdup()` in `migrator_removal_enqueue()` so no, it doesn't take the ownership. However, I see now that I don't free this value I give to `migrator_process_file()` at all.

Reply:
Rebase whacked the diff a bit but I think there is that free now (and having it fixes my const confusion too).